packages/aws-cdk-lib/aws-codepipeline/lib/private/validation.ts

```diff
@@ -81,3 +84,101 @@ function validateAgainstRegex(scope: Construct | undefined, regex: RegExp, thing
     }
   }
 }
+export function validateTriggers(gitConfiguration: GitConfiguration) {
+  const { sourceAction, pushFilter, pullRequestFilter } = gitConfiguration;
+  if (sourceAction.actionProperties.provider !== 'CodeStarSourceConnection') {
+    throw new UnscopedValidationError(`provider for actionProperties in sourceAction with name '${sourceAction.actionProperties.actionName}' must be 'CodeStarSourceConnection', got '${sourceAction.actionProperties.provider}'`);
+  }
+  if (pushFilter?.length && pullRequestFilter?.length) {
+    throw new UnscopedValidationError(`cannot specify both GitPushFilter and GitPullRequestFilter for the trigger with sourceAction with name '${sourceAction.actionProperties.actionName}'`);
+  }
+  if (!pushFilter?.length && !pullRequestFilter?.length) {
+    throw new UnscopedValidationError(`must specify either GitPushFilter or GitPullRequestFilter for the trigger with sourceAction with name '${sourceAction.actionProperties.actionName}'`);
+  }
+  if (pushFilter!== undefined) {
+    ValidateGitPushFilter(pushFilter, sourceAction.actionProperties.actionName);
+  }
+  if (pullRequestFilter!== undefined) {
+    validateGitPullRequestFilter(pullRequestFilter, sourceAction.actionProperties.actionName);
+  }
+}
+
+function ValidateGitPushFilter(pushFilter: GitPushFilter[], actionName: string) {
+  if (pushFilter.length > 3) {
+    throw new UnscopedValidationError(`length of GitPushFilter for sourceAction with name '${actionName}' must be less than or equal to 3, got ${pushFilter.length}`);
+  }
+  pushFilter.forEach(filter => {
+    validateGitFilterPropertiesLength(filter, actionName, 'GitPushFilter');
+    validateArrayLength(filter.tagsExcludes, 'tagsExcludes', actionName, 'GitPushFilter');
+    validateArrayLength(filter.tagsIncludes, 'tagsIncludes', actionName, 'GitPushFilter');
+    validateFilePathsWithBranches(filter, actionName);
+    validateTagsOrBranchesExist(filter, actionName);
+  });
+}
+
+function validateGitPullRequestFilter(pullRequestFilter: GitPullRequestFilter[], actionName: string) {
```

Comment:
Not specific to this line but all the new lines added in this file. 

It's really hard to review the validation logics to make sure we don't accidentally introduce breaking change. I appreciate the fact that we're doing synthesis validation to early error out. Curious if there is a way that we can simplify some of the validations and leave it to CFN deployment time?